lib/cli/options.js

```diff
@@ -104,24 +114,41 @@ const nargOpts = types.array
 const parse = (args = [], defaultValues = {}, ...configObjects) => {
   // save node-specific args for special handling.
   // 1. when these args have a "=" they should be considered to have values
-  // 2. if they don't, they just boolean flags
+  // 2. if they don't, they are just boolean flags
   // 3. to avoid explicitly defining the set of them, we tell yargs-parser they
   //    are ALL boolean flags.
   // 4. we can then reapply the values after yargs-parser is done.
-  const nodeArgs = (Array.isArray(args) ? args : args.split(' ')).reduce(
-    (acc, arg) => {
-      const pair = arg.split('=');
-      let flag = pair[0];
-      if (isNodeFlag(flag, false)) {
-        flag = flag.replace(/^--?/, '');
-        return arg.includes('=')
-          ? acc.concat([[flag, pair[1]]])
-          : acc.concat([[flag, true]]);
-      }
-      return acc;
-    },
-    []
-  );
+  const allArgs = Array.isArray(args) ? args : args.split(' ');
+  const nodeArgs = allArgs.reduce((acc, arg, index, allArgs) => {
+    const maybeFlag = allArgs[index - 1];
+
+    if (isNumeric(arg) && (!maybeFlag || !isMochaFlag(maybeFlag))) {
+      throw createUnsupportedError(
+        `Option ${arg} is unsupported by the mocha cli`
+      );
+    }
+    if (
+      isNumeric(arg) &&
+      isMochaFlag(maybeFlag) &&
+      expectedTypeForFlag(maybeFlag) !== 'number'
+    ) {
+      throw createInvalidArgumentTypeError(
+        `Mocha flag '--${maybeFlag}' given invalid option: '${arg}'`,
+        Number(arg),
+        expectedTypeForFlag(maybeFlag)
+      );
+    }
```

Comment:
[Feature] This check seems legitimate and good. I like it. But, it feels odd to me that it's only implemented for expected numeric types. There are booleans and strings too. I think it'd be weird as a user to only get this report for an incorrect number, but not incorrect booleans or strings.

What do you think about generalizing it to all argument types? 

I.e. instead of:

```python
if numeric and expected_type !== 'number':
  throw
```

Doing roughly:

```python
if actual_type !== expected_type:
  throw
```

Reply:
I've made a rather annoying mistake :disappointed:. I forgot to "submit" my replies to your earlier comments before requesting the second round of review :facepalm: . Really sorry about that.

I will wait for your feedback to [the previous comment](https://github.com/mochajs/mocha/pull/5263#discussion_r1869971177) before addressing this comment because I think they are related.

Reply:
Based on this and [the previous comment](https://github.com/mochajs/mocha/pull/5263#discussion_r1867727259), I've summarized the current behaviour of mocha for different flags and values data-type. 

Columns show flags and rows show values passed to flags.

Red dot 🔴 is placed on cells where mocha currently crashes without a good error message. This PR should definitely address these cases.
Orange dot 🟠 is placed where the behaviour could be improved (imo) perhaps. Mocha does not crash in these situations.

| Value / Flag | No flags  | number ⚑   | boolean ⚑ | array ⚑ |  string ⚑  |
|--------------|-----------|-------------|----------|----------| -----|
| number value  | 🔴 **Eg**: `mocha 123` - Mocha crashes with error "arg.split()"    | **Eg**: `mocha --retries 123` - 123 is assigned to `retries`   | 🔴 **Eg**: `mocha --delay 123`: `delay` is assigned value "true". 123 is treated as positional arg. Mocha crashes with error "arg.split()" | **Eg**: `mocha --file 123`: assigns value "123" to the flag. Appropriate error/warning is thrown if flag can't handle value (eg. file not found).   | **Eg**: `mocha --grep 123` assigns "123" to flag. Appropriate action is taken. |
| boolean value | **Eg**: `mocha true`: "true" is treated like a positional value for filename.    | **Eg**: `mocha --retries true`: "NaN" is assigned to flag. This value is then ignored.      | **Eg**: `mocha --delay true`: works same as `mocha --delay`. true is assigned to flag.   | **Eg** `mocha --spec true`: stringified "true" is added to array of flag values. Appropriate action is taken.   | **Eg** `mocha --ui true`. Stringified "true" is assigned to flag. |
| array value | **Eg**: `mocha v1,v2`: "v1,v2" is treated like a positional value for a filename   | 🟠 **Eg**: `mocha --retries v1,v2`  "NaN" is assigned to flag. This value is then ignored.    | **Eg**: `mocha --delay v1,v2`: value true is assigned to flag. "v1,v2" is treated like a positional filename.   | **Eg**: `mocha --spec v1,v2` Flag has value ["v1", "v2"]  | **Eg**: `mocha --grep v1,v2` String "v1,v2" is assigned to flag |
| string value | **Eg**: `mocha abc`: "abc" is treated like a positional value for a filename.    | 🟠 **Eg**: `mocha --retries abc`: "NaN" is assigned to flag. This value is then ignored.   | **Eg**: `mocha --delay abc`: value `true` is assigned to flag. "abc" is treated like positional value for a filename.   | **Eg**: `mocha --spec abc`. "abc" is added to array of flag values.   | **Eg**: `mocha --grep abc`: Works as expected. |
| No value | - | **Eg**: `mocha --retries`. Error `Not enough arguments following: retries` is thrown | **Eg**: `mocha --delay`: Flag has value `true`| **Eg**: `mocha --spec` Error `Not enough arguments following: spec` is thrown | **Eg**: `mocha grep`: Error `Not enough arguments following: grep` is thrown |


Based on the table above, I think only numeric values (or flags) seem to be causing the problem right now. That being said, I don't mind creating a more generic error handling approach. Lemme know what you prefer. I'll make the changes.

Reply:
What a great table 😄. Very helpful. I think the 🟠 ones can be ignored for now, agreed.